tfx/examples/cifar10/cifar10_utils_native_keras.py

```diff
@@ -0,0 +1,398 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Python source file includes CIFAR10 utils for Keras model.
+
+The utilities in this file are used to build a model with native Keras.
+This module file will be used in Transform and generic Trainer.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+from typing import List, Text
+import absl
+import tensorflow as tf
+import tensorflow_transform as tft
+
+import flatbuffers
+# pylint: disable=g-direct-tensorflow-import
+from tflite_support import metadata as _metadata
+from tflite_support import metadata_schema_py_generated as _metadata_fb
+# pylint: enable=g-direct-tensorflow-import
+
+from tfx.components.trainer.rewriting import converters
+from tfx.components.trainer.rewriting import rewriter
+from tfx.components.trainer.rewriting import rewriter_factory
+
+from tfx.components.trainer.executor import TrainerFnArgs
+
+# When training on the whole dataset use following constants instead.
+# This setting should give ~91% accuracy on the whole test set
+# _TRAIN_DATA_SIZE = 50000
+# _EVAL_DATA_SIZE = 10000
+# _TRAIN_BATCH_SIZE = 64
+# _EVAL_BATCH_SIZE = 64
+# _CLASSIFIER_LEARNING_RATE = 3e-4
+# _FINETUNE_LEARNING_RATE = 5e-5
+# _CLASSIFIER_EPOCHS = 12
+
+_TRAIN_DATA_SIZE = 128
+_EVAL_DATA_SIZE = 128
+_TRAIN_BATCH_SIZE = 32
+_EVAL_BATCH_SIZE = 32
+_CLASSIFIER_LEARNING_RATE = 1e-3
+_FINETUNE_LEARNING_RATE = 7e-6
+_CLASSIFIER_EPOCHS = 30
+
+_IMAGE_KEY = 'image'
+_LABEL_KEY = 'label'
+
+_TFLITE_MODEL_NAME = 'tflite'
+_LABEL_MAP_FILE_PATH = 'cifar10/data/labels.txt'
+
+def _transformed_name(key):
+  return key + '_xf'
+
+def _gzip_reader_fn(filenames):
+  """Small utility returning a record reader that can read gzip'ed files."""
+  return tf.data.TFRecordDataset(filenames, compression_type='GZIP')
+
+def _get_serve_image_fn(model):
+  """Returns a function that feeds the input tensor into the model."""
+
+  @tf.function
+  def serve_image_fn(image_tensor):
+    """Returns the output to be used in the serving signature.
+
+    Args:
+      image_tensor: A tensor represeting input image. The image should
+        have 3 channels.
+
+    Returns:
+      The model's predicton on input image tensor
+    """
+    return model(image_tensor)
+
+  return serve_image_fn
+
+def _image_augmentation(image_features):
+  """Perform image augmentation on batches of images .
+
+  Args:
+    image_feature: a batch of image features
+
+  Returns:
+    The augmented image features
+  """
+  batch_size = tf.shape(image_features)[0]
+  image_features = tf.image.random_flip_left_right(image_features)
+  image_features = tf.image.resize_with_crop_or_pad(image_features, 250, 250)
+  image_features = tf.image.random_crop(image_features,
+                                        (batch_size, 224, 224, 3))
+  return image_features
+
+def _data_augmentation(feature_dict):
```

Comment:
why not do this in tft?

Reply:
The data augmentation is a special type of pre-processing: it aims to add randomness to the dataset and should be applied on the fly during training, i.e. the model should see a slightly different dataset across different epochs during training. If we put this in Transform, this data augmentation will only be applied once for the whole dataset, and the model will see a consistent dataset throughout training, which loses the point of data augmentation. This is the workaround after discussing with David, and we are working a design doc for different pre-processing operation types. We hope to initiate a discussion on  the best way to handle these special pre-processing operations.

Reply:
sg, could you add some comments here?

Reply:
Sure I added some comments in the input_fn, at the location where the data augmentation is applied.